File: src/image/jpeg-handler.ts
import type { PdfRef } from '../core/types.js';
import type { ObjectStore } from '../core/object-store.js';
import { pdfStream, pdfName, pdfNum, pdfArray } from '../core/objects.js';
import type { ImageInfo } from './image-utils.js';

/**
 * Read a 16-bit big-endian unsigned integer from a Uint8Array.
 */
function readUint16BE(data: Uint8Array, offset: number): number {
  return (data[offset] << 8) | data[offset + 1];
}

/**
 * Parse a JPEG file to extract image dimensions, color space, and bit depth.
 *
 * JPEG structure:
 * - File starts with SOI marker (FF D8)
 * - Followed by segments, each starting with FF xx (marker)
 * - Most segments have a 2-byte length field after the marker
 * - SOF0 (FF C0) and SOF2 (FF C2) contain image parameters:
 *   - Byte 0-1: marker (FF C0/C2)
 *   - Byte 2-3: segment length
 *   - Byte 4: precision (bits per component)
 *   - Byte 5-6: height
 *   - Byte 7-8: width
 *   - Byte 9: number of components (1=Gray, 3=RGB/YCbCr, 4=CMYK)
 */
export function parseJpeg(data: Uint8Array): ImageInfo {
  if (data.length < 2 || data[0] !== 0xff || data[1] !== 0xd8) {
    throw new Error('Invalid JPEG: missing SOI marker');
  }

  let offset = 2;

  while (offset < data.length - 1) {
    // Find next marker
    if (data[offset] !== 0xff) {
      offset++;
      continue;
    }

    // Skip padding FF bytes
    while (offset < data.length && data[offset] === 0xff) {
      offset++;
    }

    if (offset >= data.length) break;

    const marker = data[offset];
    offset++;

    // Check for SOF markers (Start Of Frame)
    // SOF0 (0xC0) = Baseline DCT
    // SOF1 (0xC1) = Extended sequential DCT
    // SOF2 (0xC2) = Progressive DCT
    // SOF3 (0xC3) = Lossless
    // SOF5-SOF7, SOF9-SOF11, SOF13-SOF15 are also SOF markers
    const isSOF =
      (marker >= 0xc0 && marker <= 0xc3) ||
      (marker >= 0xc5 && marker <= 0xc7) ||
      (marker >= 0xc9 && marker <= 0xcb) ||
      (marker >= 0xcd && marker <= 0xcf);

    if (isSOF) {
      if (offset + 7 > data.length) {
        throw new Error('Invalid JPEG: truncated SOF segment');
      }

      // Skip segment length (2 bytes)
      const precision = data[offset + 2];
      const height = readUint16BE(data, offset + 3);
      const width = readUint16BE(data, offset + 5);
      const numComponents = data[offset + 7];

      let colorSpace: 'DeviceRGB' | 'DeviceGray' | 'DeviceCMYK';
      switch (numComponents) {
        case 1:
          colorSpace = 'DeviceGray';
          break;
        case 3:
          colorSpace = 'DeviceRGB';
          break;
        case 4:
          colorSpace = 'DeviceCMYK';
          break;
        default:
          throw new Error(`Unsupported JPEG component count: ${numComponents}`);
      }

      return {
        width,
        height,
        colorSpace,
        bitsPerComponent: precision,
        hasAlpha: false, // JPEG never has alpha
      };
    }

    // For other markers, skip to next segment
    // Markers with no payload:
    if (marker === 0xd9) {
      // EOI (End Of Image)
      break;
    }
    if (marker === 0x00 || marker === 0x01 || (marker >= 0xd0 && marker <= 0xd7)) {
      // Standalone markers (RST0-RST7, TEM) - no payload
      continue;
    }

    // Markers with payload: read length
    if (offset + 1 >= data.length) break;
    const segmentLength = readUint16BE(data, offset);
    offset += segmentLength;
  }

  throw new Error('Invalid JPEG: no SOF marker found');
}

/**
 * Embed a JPEG image into a PDF ObjectStore.
 * JPEG data can be embedded directly as a DCTDecode stream.
 */
export function embedJpeg(
  store: ObjectStore,
  data: Uint8Array,
  info: ImageInfo,
): PdfRef {
  const ref = store.allocRef();

  const streamDict: Record<string, import('../core/types.js').PdfObject> = {
    Type: pdfName('XObject'),
    Subtype: pdfName('Image'),
    Width: pdfNum(info.width),
    Height: pdfNum(info.height),
    ColorSpace: pdfName(info.colorSpace),
    BitsPerComponent: pdfNum(info.bitsPerComponent),
    Filter: pdfName('DCTDecode'),
    Length: pdfNum(data.length),
  };

  // For CMYK JPEGs, add a Decode array to invert the values
  // since JPEG stores CMYK as inverted (YCCK)
  if (info.colorSpace === 'DeviceCMYK') {
    streamDict['Decode'] = pdfArray(
      pdfNum(1), pdfNum(0),
      pdfNum(1), pdfNum(0),
      pdfNum(1), pdfNum(0),
      pdfNum(1), pdfNum(0),
    );
  }

  const stream = pdfStream(streamDict, data);
  store.set(ref, stream);

  return ref;
}

File: src/image/jpeg_handler.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const core = @import("../core/types.zig");
const ObjectStore = @import("../core/object_store.zig").ObjectStore;
const Ref = core.Ref;
const PdfObject = core.PdfObject;
const image_embedder = @import("image_embedder.zig");
const ImageHandle = image_embedder.ImageHandle;
const ImageFormat = image_embedder.ImageFormat;

/// Color space types found in JPEG files.
pub const ColorSpace = enum {
    device_gray,
    device_rgb,
    device_cmyk,

    /// Returns the PDF name for this color space.
    pub fn pdfName(self: ColorSpace) []const u8 {
        return switch (self) {
            .device_gray => "DeviceGray",
            .device_rgb => "DeviceRGB",
            .device_cmyk => "DeviceCMYK",
        };
    }
};

/// Information extracted from a JPEG header.
pub const JpegInfo = struct {
    width: u32,
    height: u32,
    components: u8,
    bits_per_component: u8,
    color_space: ColorSpace,
};

pub const JpegError = error{
    InvalidJpeg,
    UnsupportedJpeg,
    UnexpectedEndOfData,
};

/// Parse a JPEG header to extract dimensions and color information.
/// Reads the SOI marker (0xFFD8) and scans for SOF0 (0xFFC0) or SOF2 (0xFFC2)
/// markers to extract width, height, and component count.
pub fn parseJpeg(data: []const u8) JpegError!JpegInfo {
    if (data.len < 2) return JpegError.InvalidJpeg;

    // Verify SOI marker
    if (data[0] != 0xFF or data[1] != 0xD8) return JpegError.InvalidJpeg;

    var offset: usize = 2;

    while (offset + 1 < data.len) {
        // Find next marker
        if (data[offset] != 0xFF) return JpegError.InvalidJpeg;

        // Skip padding 0xFF bytes
        while (offset + 1 < data.len and data[offset + 1] == 0xFF) {
            offset += 1;
        }

        if (offset + 1 >= data.len) return JpegError.UnexpectedEndOfData;

        const marker = data[offset + 1];
        offset += 2;

        // SOF0 (baseline) or SOF2 (progressive)
        if (marker == 0xC0 or marker == 0xC2) {
            if (offset + 8 > data.len) return JpegError.UnexpectedEndOfData;

            const bits_per_component = data[offset + 2];
            const height = @as(u32, data[offset + 3]) << 8 | @as(u32, data[offset + 4]);
            const width = @as(u32, data[offset + 5]) << 8 | @as(u32, data[offset + 6]);
            const components = data[offset + 7];

            const color_space: ColorSpace = switch (components) {
                1 => .device_gray,
                3 => .device_rgb,
                4 => .device_cmyk,
                else => return JpegError.UnsupportedJpeg,
            };

            return JpegInfo{
                .width = width,
                .height = height,
                .components = components,
                .bits_per_component = bits_per_component,
                .color_space = color_space,
            };
        }

        // EOI marker
        if (marker == 0xD9) return JpegError.InvalidJpeg;

        // SOS marker - data follows, no SOF found before SOS
        if (marker == 0xDA) return JpegError.InvalidJpeg;

        // Skip marker segment (read length and advance)
        if (offset + 2 > data.len) return JpegError.UnexpectedEndOfData;
        const seg_len = @as(usize, data[offset]) << 8 | @as(usize, data[offset + 1]);
        if (seg_len < 2) return JpegError.InvalidJpeg;
        offset += seg_len;
    }

    return JpegError.InvalidJpeg;
}

/// Create a PDF image XObject for a JPEG image.
/// The raw JPEG data is stored as the stream with /DCTDecode filter.
pub fn embedJpeg(allocator: Allocator, store: *ObjectStore, data: []const u8) !ImageHandle {
    const info = try parseJpeg(data);

    const ref = try store.allocate();

    var dict: std.StringHashMapUnmanaged(PdfObject) = .{};
    try dict.put(allocator, "Type", core.pdfName("XObject"));
    try dict.put(allocator, "Subtype", core.pdfName("Image"));
    try dict.put(allocator, "Width", core.pdfInt(@intCast(info.width)));
    try dict.put(allocator, "Height", core.pdfInt(@intCast(info.height)));
    try dict.put(allocator, "ColorSpace", core.pdfName(info.color_space.pdfName()));
    try dict.put(allocator, "BitsPerComponent", core.pdfInt(@intCast(info.bits_per_component)));
    try dict.put(allocator, "Filter", core.pdfName("DCTDecode"));
    try dict.put(allocator, "Length", core.pdfInt(@intCast(data.len)));

    const stream_obj = PdfObject{ .stream_obj = .{
        .dict = dict,
        .data = data,
    } };

    store.put(ref, stream_obj);

    return ImageHandle{
        .ref = ref,
        .width = info.width,
        .height = info.height,
        .format = .jpeg,
    };
}

// -- Tests --

test "parseJpeg: invalid data" {
    const result = parseJpeg(&[_]u8{ 0x00, 0x00 });
    try std.testing.expectError(JpegError.InvalidJpeg, result);
}

test "parseJpeg: too short" {
    const result = parseJpeg(&[_]u8{0xFF});
    try std.testing.expectError(JpegError.InvalidJpeg, result);
}

test "parseJpeg: valid minimal JPEG with SOF0" {
    // SOI + APP0 marker (minimal) + SOF0 marker
    const data = [_]u8{
        0xFF, 0xD8, // SOI
        0xFF, 0xE0, // APP0
        0x00, 0x02, // length = 2 (just the length field)
        0xFF, 0xC0, // SOF0
        0x00, 0x0B, // length = 11
        0x08, // bits per component
        0x00, 0x40, // height = 64
        0x00, 0x80, // width = 128
        0x03, // 3 components (RGB)
        0x01, 0x11, 0x00, // component 1
        0x02, 0x11, 0x01, // component 2
        0x03, 0x11, 0x01, // component 3
    };

    const info = try parseJpeg(&data);
    try std.testing.expectEqual(@as(u32, 128), info.width);
    try std.testing.expectEqual(@as(u32, 64), info.height);
    try std.testing.expectEqual(@as(u8, 3), info.components);
    try std.testing.expectEqual(@as(u8, 8), info.bits_per_component);
    try std.testing.expectEqual(ColorSpace.device_rgb, info.color_space);
}

test "parseJpeg: grayscale" {
    const data = [_]u8{
        0xFF, 0xD8, // SOI
        0xFF, 0xC0, // SOF0
        0x00, 0x08, // length = 8
        0x08, // bits per component
        0x00, 0x10, // height = 16
        0x00, 0x20, // width = 32
        0x01, // 1 component (grayscale)
        0x01, 0x11, 0x00, // component 1
    };

    const info = try parseJpeg(&data);
    try std.testing.expectEqual(@as(u32, 32), info.width);
    try std.testing.expectEqual(@as(u32, 16), info.height);
    try std.testing.expectEqual(ColorSpace.device_gray, info.color_space);
}
